run_test.sh

```diff
@@ -32,36 +38,48 @@ find generated -type f -not \( -name "*.expected" -or -name "__init__.py" \) -de
 )
 
 (
-    # Run mypy (always under python3)
+    # Run mypy
+    eval "$(pyenv init -)"
+    pyenv shell $PY_VER_MYPY
+    PY_VERSION=`python -c 'import sys; print(sys.version.split()[0])'`
+    VENV=venv_$PY_VERSION
 
     # Create virtualenv
-    if [[ -z $SKIP_CLEAN ]] || [[ ! -e $MYPY_VENV ]]; then
+    if [[ -z $SKIP_CLEAN ]] || [[ ! -e $VENV ]]; then
         python3 --version
         python3 -m pip --version
-        python3 -m virtualenv $MYPY_VENV
+        python3 -m virtualenv $VENV
     fi
-    source $MYPY_VENV/bin/activate
-    if [[ -z $SKIP_CLEAN ]] || [[ ! -e $MYPY_VENV ]]; then
+    source $VENV/bin/activate
+    if [[ -z $SKIP_CLEAN ]] || [[ ! -e $VENV ]]; then
         python3 -m pip install setuptools
         python3 -m pip install git+https://github.com/python/mypy.git@985a20d87eb3a516ff4457041a77026b4c6bd784
     fi
 
     # Run mypy
-    for PY in 2.7 3.5; do
-        mypy --custom-typeshed-dir=$CUSTOM_TYPESHED_DIR --python-version=$PY --pretty --show-error-codes python/mypy_protobuf.py test/ generated/
-        if ! diff <(mypy --custom-typeshed-dir=$CUSTOM_TYPESHED_DIR --python-version=$PY python/mypy_protobuf.py test_negative/ generated/) test_negative/output.expected.$PY; then
-            echo -e "${RED}test_negative/output.expected.$PY didnt match. Copying over for you. Now rerun${NC}"
-            for PY in 2.7 3.5; do
-                mypy --custom-typeshed-dir=$CUSTOM_TYPESHED_DIR --python-version=$PY python/mypy_protobuf.py test_negative/ generated/ > test_negative/output.expected.$PY || true
-            done
-            exit 1
-        fi
-    done
+    mypy --custom-typeshed-dir=$CUSTOM_TYPESHED_DIR --python-version=$PY_VER_MYPY_TARGET --pretty --show-error-codes python/mypy_protobuf.py test/ generated/
+    if ! diff <(mypy --custom-typeshed-dir=$CUSTOM_TYPESHED_DIR --python-version=$PY_VER_MYPY_TARGET python/mypy_protobuf.py test_negative/ generated/) test_negative/output.expected.$PY_VER_MYPY_TARGET; then
+        echo -e "${RED}test_negative/output.expected.$PY_VER_MYPY_TARGET didnt match. Copying over for you. Now rerun${NC}"
+        for PY_VER_MYPY_TARGET in 2.7 3.5; do
```

Comment:
this loop is not necessary because we need to run tests twice anyway to check both targets

Reply:
yep - that makes sense, will update. Thanks